### DISCUSSION TITLE
Notes for August 31st, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#6440] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-08-31.md

1. Short term work triage
    1. Partial properties
    2. Using aliases for any type
    3. Null-conditional assignment
    4. `await?`
    5. `params IEnumerable`
    6. Discriminated Unions
    7. Roles
    8. `pragma warning enable`
    9. Ignored directives support
    10. Primary constructors
    11. Switch expression as a statement

**Comment by TahirAhmadov** [comment]
I'm excited for `await?`. 
Also, for `params IEnumerable<T>`, can we at least come to a clear understanding of what needs to be supported in a perfect world? Meaning, if we forget the existence of `params T[]` for a moment, and could add `params Span<T>` support to the runtime, what would be the ideal solution? I'm a little unclear on this because `IEnumerable<T>` and `Span<T>` are not compatible; the way I see it, it will be necessary to define both overloads. Basically - million bucks question - can we create one `params` to rule them all? One which would take `IEnumerable<T>` and `Span<T>` impicitly?

**Reply by HaloFour** [reply]
> and could add params Span<T> support to the runtime, what would be the ideal solution?

IIRC no runtime changes are necessary here.  Whether a parameter is `params` is just a bit of metadata that the compiler interprets when calling the method.

> can we create one `params` to rule them all? One which would take `IEnumerable<T>` and `Span<T>` impicitly?

I'd kind of doubt it.  Those two types have very different rules especially around scope and lifecycle that can't be handwaved away.

**Reply by TahirAhmadov** [reply]
Thinking about it, it seems like 2 new special `ref struct`s would have to be created, which would essentially wrap around either `Span` or `IEnumerable`; difference being, one provides `this[index]` and `Count` and one doesn't, but both are enumerable somehow.
PS. Then each method can decide whether it needs just "enumerable" struct, or whether it needs indexing and count, as well.
PPS. Obviously, big problem - all public API of said structs will incur an overhead of some sort, due to accessing one of the 2 options - created from Span or IEnumerable.
PPPS. Or maybe even 3 struct types, to mirror the separation between IEnumerable/ICollection/IList?

**Reply by TahirAhmadov** [reply]
> IIRC no runtime changes are necessary here. Whether a parameter is params is just a bit of metadata that the compiler interprets when calling the method.

I was thinking about the conversation somewhere about the stack allocation of stuff and releasing it, and how that needs runtime support.

**Reply by 333fred** [reply]
These questions are why it's not on the shortlist 😉

**Reply by HaloFour** [reply]
I think that's more a question of how the compiler emits the IL, as a `stackalloc` in a loop can result in problems.  But I don't think that's specifically a problem for `params Span<T>`, and I'd assume naively that the compiler could mitigate that by preallocating a buffer on the stack and reuse it in those circumstances.  But perhaps the runtime support to deallocate stack space would be valuable on its own.

**Reply by HaloFour** [reply]
@333fred 

> These questions are why it's not on the shortlist 😉

What's the fun in trying to solve the easy problems?  😀

**Reply by TahirAhmadov** [reply]
@333fred I guess my request was, can the discussion about this be prioritized a little? Can we treat this not as a small feature, but a medium-sized one, (including both IEnumerable and Span) and start the process accordingly?

**Reply by 333fred** [reply]
When we come back from break, we will be triaging our list of issues.

**Reply by MgSam** [reply]
Putting primary constructors, probably the most debated C# feature ever, on the "shortlist" is hilarious.

**Reply by HaloFour** [reply]
Primary constructors are already in the language, it makes sense to explore how they might be supported with non-records.  The "debate" that arose during the CodePlex days was largely around trying to make them have parity with normal constructors by allowing raw code statements to appear within the body of the method.  IMO none of that is necessary to still bring benefits to the language.

**Reply by 333fred** [reply]
@MgSam it is specifically on the _prototyping_ shortlist because we feel that doing so will help drive the design.

**Reply by MgSam** [reply]
> IMO none of that is necessary to still bring benefits to the language.

I'm sure others will disagree. 🙂

**Reply by HaloFour** [reply]
@MgSam 

> I'm sure others will disagree. 🙂

Sure, I doubt any design decisions have ever been unanimous.  That is my opinion, though, based on my experience with working with other languages that have these constructs where they are usually opinionated to a degree.  They don't have to achieve parity because you always have the escape hatch of defining a full constructor.  The question is how many use cases you can cover before you hit a cliff edge, and how steep that is.

**Reply by HaloFour** [reply]
Admittedly I think most of the benefits of primary constructors for non-record types could be achieved via source generators.  To borrow a page from [Lombok](https://projectlombok.org/features/constructor):

```cs
[RequiredArgsConstructor]
public partial class Person {
    private readonly string name;
    private readonly int age;
}

// generated:
public partial class Person {
    public Person(string name, int age) {
        this.name = name;
        this.age = age;
    }
}
```
